### ISSUE TITLE
Can Cryptree be implemented on an MST or Prolly tree?

### ISSUE THREAD

**Issue by Nuhvi** [Peergos/Peergos#1049]
Is there anything in the Cryptree design that either require Hamts or just negates the benefits from ordered trees like [MST](https://inria.hal.science/hal-02303490/document) and [Prolly trees](https://github.com/attic-labs/noms/blob/master/doc/intro.md#prolly-trees-probabilistic-b-trees)?

**Comment by ianopolous** [comment]
The only critical assumptions we make are the following. 
1) It is semantically a Map
2) Removals from the map remove the underlying blocks (i.e. it's not append only)

We additionally assume 
3) the merkle root is insertion order independent for a given set of mappings. This helps with merging concurrent edits automatically. This is true for CHAMPs but not strictly necessary. We were originally using a merkle-btree, but switched because it lacked this property.

You could drop (2) as well but then access revocation is not as effective and you're probably not GDPR compliant. 

The keys are basically random though, so an ordered tree wouldn't give any benefit.

**Comment by Nuhvi** [comment]
Why are keys random? because of encryption?

I wonder if Public drive should be in an ordered tree, for performance. I understand that this makes converting subfolders to private will be slower, but how often does that happen?

Also, how often do you see apps built on top of Peergos use folders semantics and granular access control, as opposed to a flat bucket of files with all-or-nothing access to that bucket?

Because if it is the latter, I imagine it would be easier to create a PDS with repos made of ordered trees, where nodes are either encrypted or not, and if encrypted, they can use a simple scheme of a single symmetric key. Something akin to buckets symantics in S3.

**Comment by ianopolous** [comment]
Keys are random for privacy. The server can't deduce which mappings belong to the same file or directory, or even if one is a file or directory. Each mapping (chunk) is max 5 MiB. 

For public files we currently use a totally separate champ, nothing to do with cryptree, which is an inode filesystem mapping paths to capabilities. This design was mainly easy to implement, we're not especially wedded to it. Longer term I could see us adopting iroh for public data.

Apps built on peergos don't need to understand or implement encryption. They are basically standard web apps, with a REST api to get files/dirs etc. As far as access goes, cryptree is recursive. So a single dir read/write cap grants you access to the entire subtree.

**Comment by ianopolous** [comment]
The typical argument decentralization people make is that, "oh if people want more privacy they can just self host". But most people won't be able to self host, so we wanted a design that enables the average non technical user to still get first-in-class privacy, even from their host. Combined with portable accounts and content addressing this is what we want from the web, a built-in defense against surveillance capitalism, and actually usable by non technical people.

**Comment by ianopolous** [comment]
Scanning can still be done, but client side. If you try the search functionality in Peergos this shows the current speed of it without any optimisation (It was fast enough for us to not bother optimising it yet). For an example in the old UI see https://peergos.org/posts/file-search

Champs and cryptree are pretty simple. Our champ implementation is 800 lines, and cryptree is 1300 lines, each in a single file. 

Also, if clients are caching most of the blocks locally, then the client-side search can be just as fast as a server-side search.

**Comment by Nuhvi** [comment]
If keys are random though, so lookup is O(n) of items in a folder right?

**Comment by ianopolous** [comment]
The caps in a folder are named. so name based search doesn't need to download any metadata for the files at all. If you want to search based on other metadata like modification time etc. then yes. 

There is actually also some flexibility here. We don't enforce any ordering of the entries in a directory. But we could.

**Comment by Nuhvi** [comment]
Yes but if I have to download the names of all files in a folder, that still means I am downloading O(n) names to find what I am looking for. whereas in an ordered tree, you will need O(k * log n) where n is the number of files, and k is the branching factor.

**Comment by Nuhvi** [comment]
If the goal is a Dropbox alternative, but with sovereignty and encryption at rest, I agree Cryptree is great, and maybe that is all we will ever need. But the idea of PDS always struck me to be closer to a backend and a database, not just storage.

Thanks for the answers.

**Comment by ianopolous** [comment]
I think with a filesystem foundation you can always build a DB on top. It's the database layer that solves for the kind of thing you sound like you want. The only reason we haven't gone down that route yet is we haven't hit a limit for the use cases we've explored in various kinds of apps and social media. With a clean filesystem layer, each app can choose its own database, if it even needs one.

**Comment by Nuhvi** [comment]
I am not sure about that. seems to me, that an authenticated B-tree, is the lowest level of abstraction, on top of which a filesystem can be built, but the metadata of files, and the lack of ordered keys in Cryptrees, would make building a performant indexable database out of that filesystem... quite hard.

That is why I am considering MST and Prolly tree, not because I don't want a filesystem or at least an interface with Fuse, but because it is simpler and lower level. I can imagine how to use a k-v store to add the granular access control of Cryptree while loosing some performance, but I can't imagine how to use CHAMP and Cryptree, to make a database that is as performant as an MST.

Maybe you mean one file databases like SQLite, but I won't even start imagining how the sync of that works, but maybe it does? I feel that it at least will require full-sync all the time, and I really aspire for partial sync, because clients have limited storage and data subscription budget.

**Comment by ianopolous** [comment]
Real databases are themselves built on top of filesystems with similar properties. That's where I'm coming from. 

We already have partial file sync. If you modify a file you only need to upload the modified chunk, not the entire file.

**Comment by Nuhvi** [comment]
That is not what I meant by partial sync. Sure if only a part of the file changes you can get the diff, but I mean if the database is a signle SQLlite file, then you simply can't sync with one table, or a range of rows, yes you can get a range of the file, but you won't know which range do yo need, because it is opaque binary until you download some of it to learn more.

And... maybe SQL makes that easy, maybe you can read first X bytes and learn where to look next, but I am not sure, seems to me that a balanced merkle tree is a more general approach. But I can be wrong.

My anecdotes are Ethereum and Farcaster using Patricia Merkle trie, Bluesky using MST, Dolthub using prolly trees, Hyperbee wich encodes an append-only b-tree in a hypercore, instead of implementing an SQLlite in a hypercore, and the list goes on.

It is worth testing though, see if you can build a performant local-first app (maybe even multiplayer) on a single SQL file in Peergos

**Comment by Nuhvi** [comment]
One disadvantage of the SQLite file, is that it is not clear what the would the strong link and the proof of a certain entry in the database look like. (blake3 would definitely help here though since it will merklize the sqlite file itself).

but strong links that look like `gitfordata:<ed25519 pubkey>/<bucket>/<key>?commit=<hash>` is nicer than the alternative.

I also suspect that, with most users being mobile users instead of desktop users nowadays, the bucket + tagging semantics might have actually become the more intuitive paradigm than the filesystem.
